crates/egui/src/containers/popup.rs

```diff
@@ -1,457 +1,517 @@
-
-use pass_state::PerWidgetTooltipState;
-
 use crate::{
-    pass_state, vec2, AboveOrBelow, Align, Align2, Area, AreaState, Context, Frame, Id,
-    InnerResponse, Key, LayerId, Layout, Order, Pos2, Rect, Response, Sense, Ui, UiKind, Vec2,
-    Widget, WidgetText,
+    Area, AreaState, Context, Frame, Id, InnerResponse, Key, LayerId, Layout, Order, Response,
+    Sense, Ui, UiKind,
 };
+use emath::{vec2, Align, Align4, Pos2, Rect};
+use std::iter::once;
 
-// ----------------------------------------------------------------------------
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PopupAnchor {
+    /// Show the popup at the given rect.
+    Rect(Rect),
 
-fn when_was_a_toolip_last_shown_id() -> Id {
-    Id::new("when_was_a_toolip_last_shown")
-}
+    /// Show the popup at the current mouse pointer position.
+    Pointer,
 
-pub fn seconds_since_last_tooltip(ctx: &Context) -> f32 {
-    let when_was_a_toolip_last_shown =
-        ctx.data(|d| d.get_temp::<f64>(when_was_a_toolip_last_shown_id()));
+    /// Show the popup at the mouse pointer and remember the position (like a context menu).
+    PointerFixed,
 
-    if let Some(when_was_a_toolip_last_shown) = when_was_a_toolip_last_shown {
-        let now = ctx.input(|i| i.time);
-        (now - when_was_a_toolip_last_shown) as f32
-    } else {
-        f32::INFINITY
+    /// Show the popup at the given position.
+    Position(Pos2),
+}
+
+impl From<Rect> for PopupAnchor {
+    fn from(rect: Rect) -> Self {
+        Self::Rect(rect)
     }
 }
 
-fn remember_that_tooltip_was_shown(ctx: &Context) {
-    let now = ctx.input(|i| i.time);
-    ctx.data_mut(|data| data.insert_temp::<f64>(when_was_a_toolip_last_shown_id(), now));
+impl From<Pos2> for PopupAnchor {
+    fn from(pos: Pos2) -> Self {
+        Self::Position(pos)
+    }
 }
 
-// ----------------------------------------------------------------------------
-
-/// Show a tooltip at the current pointer position (if any).
-///
-/// Most of the time it is easier to use [`Response::on_hover_ui`].
-///
-/// See also [`show_tooltip_text`].
-///
-/// Returns `None` if the tooltip could not be placed.
-///
-/// ```
-/// # egui::__run_test_ui(|ui| {
-/// if ui.ui_contains_pointer() {
-///     egui::show_tooltip(ui.ctx(), ui.layer_id(), egui::Id::new("my_tooltip"), |ui| {
-///         ui.label("Helpful text");
-///     });
-/// }
-/// # });
-/// ```
-pub fn show_tooltip<R>(
-    ctx: &Context,
-    parent_layer: LayerId,
-    widget_id: Id,
-    add_contents: impl FnOnce(&mut Ui) -> R,
-) -> Option<R> {
-    show_tooltip_at_pointer(ctx, parent_layer, widget_id, add_contents)
+impl From<&Response> for PopupAnchor {
+    fn from(response: &Response) -> Self {
+        let mut widget_rect = response.rect;
+        if let Some(to_global) = response.ctx.layer_transform_to_global(response.layer_id) {
+            widget_rect = to_global * widget_rect;
+        }
+        Self::Rect(widget_rect)
```

Comment:
Oh, it's not the rectangle of the popup, but the rectangle of the parent widget _next_ to which the popup should be shown. That should be made clear in the docstring.